bin/kano-updater-recovery

```diff
@@ -0,0 +1,351 @@
+
+
+def reset_wifi_retries():
+    """
+    Resets the count of recovery mechanism has encountered the WiFi
+    connection failure screen.
+    """
+
+    try:
+        import os
+        os.remove(RECOVERY_STATUS_PATH)
+    except (IOError, OSError):
+        pass
+
+
+def is_err_being_shown():
+    """Check if a UFx error splash is currently being shown.
+
+    Returns:
+        bool: Whether a recovery-err-splash service is active
+    """
+
+    import os
+    return (os.system('systemctl is-active -q recovery-err-splash@*.service') == 0)
+
+
+def run_updater():
+    """Launch the updater recovery in the background
+
+    Assumes the state was already checked to be INSTALLING_UPDATES.
+
+    Returns:
+        subprocess.Popen: Launched updater process
+    """
+
+    import os
+    import subprocess
+
+    return subprocess.Popen(
+        ['kano-updater', 'install'],
+        shell=False,
+        preexec_fn=os.setsid
+    )
+
+
+def stop_updater(proc):
+    """Stops the Updater process
+
+    Args:
+        proc (subprocess.Popen): The process of the Updater
+    """
+
+    import os
+    import signal
+    os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
+
+
+def parse_updater_rc(rc):
+    """
+    Parses the return code from the updater terminating to the language of the
+    UFx codes.
+
+    Args:
+        rc (int): Return code from the updater's termination
+
+    Returns:
+        int: Return code corresponding to the updater's finished state (Member
+        of :class:`RC` class).
+    """
+
+    from kano_updater.return_codes import RC as UpdaterRC
+
+    if rc == UpdaterRC.SUCCESS:
+        reset_wifi_retries()
+        return rc
+
+    elif rc in [UpdaterRC.NO_NETWORK, UpdaterRC.CANNOT_REACH_KANO]:
+        if get_wifi_retries() > PERMITTED_WIFI_RETRIES:
+            return RC.UF4
+        else:
+            increment_wifi_retries()
+            return RC.UF3
+
+    elif rc == UpdaterRC.HANGED_INDEFINITELY:
+        return RC.UF5
+
+    else:
+        return RC.UF2
+
+
+def reboot(delay=0):
+    """Perform a system reboot.
+
+    Args:
+        delay (int): Time in seconds to delay the operation (same thread).
+    """
+
+    if delay:
+        import time
+        time.sleep(delay)
+
+    import os
+    os.system('systemctl reboot')
+
+
+def prepare_for_hard_shutdown():
+    """Configure the system to handle a hard poweroff gracefully."""
+
+    pass  # TODO: Insert black magic here.
+
+
+def infinite_loop():
+    """Start an infinite loop within the main thread.
+
+    Note:
+        This function is designed to be the last thing you call. It will
+        block indefinitely.
+    """
+
+    import time
+    while True:
+        time.sleep(1 * HOUR)
+
+
+def main():
+    """Main recovery program."""
+
+    from kano_updater.status import UpdaterStatus
+    if not UpdaterStatus.get_instance().is_recovery_needed():
+        return RC.SUCCESS
+
+    # Check if any UFx is being shown, e.g. kano-os-recovery had a timeout.
```

Comment:
I am guessing we are preparing for a shutdown initiated by an external process? Can we briefly explain this?

Reply:
So when there is an error during the entire flow (split between `kano-os-recovery` and `kano-updater-recovery`) the spec is to show a `recovery-err-splash` indefinitely. This means that the user is expected to unplug and (not) try again depending on the error code shown.

Here, we check whether `kano-os-recovery` has failed to quit `kano-updater-recovery` early. Updated the comment.